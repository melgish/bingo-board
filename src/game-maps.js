export const coverAll = [
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
]

export const bigBox = [
  [1, 1, 1, 1, 1],
  [1, 0, 0, 0, 1],
  [1, 0, 0, 0, 1],
  [1, 0, 0, 0, 1],
  [1, 1, 1, 1, 1],
]

export const smallBox = [
  [0, 0, 0, 0, 0],
  [0, 1, 1, 1, 0],
  [0, 1, 0, 1, 0],
  [0, 1, 1, 1, 0],
  [0, 0, 0, 0, 0],
]

export const bigCorners = [
  [1, 1, 0, 1, 1],
  [1, 1, 0, 1, 1],
  [0, 0, 0, 0, 0],
  [1, 1, 0, 1, 1],
  [1, 1, 0, 1, 1],
]

export const letterX = [
  [1, 0, 0, 0, 1],
  [0, 1, 0, 1, 0],
  [0, 0, 1, 0, 0],
  [0, 1, 0, 1, 0],
  [1, 0, 0, 0, 1],
]

export const letterT = [
  [1, 1, 1, 1, 1],
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
]

export const cross = [
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
  [1, 1, 1, 1, 1],
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
]

export const letterL = [
  [1, 0, 0, 0, 0],
  [1, 0, 0, 0, 0],
  [1, 0, 0, 0, 0],
  [1, 0, 0, 0, 0],
  [1, 1, 1, 1, 1],
]

export const normals = [
  [
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
  ],
  [
    [0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
  ],
  [
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
  ],
  [
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
  ],
  [
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
  ],
  [
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
  ],
  [
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
  ],
  [
    [0, 0, 1, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 1, 0, 0],
  ],
  [
    [0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0],
  ],
  [
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 1],
  ],
  [
    [1, 0, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 0, 1],
  ],
  [
    [0, 0, 0, 0, 1],
    [0, 0, 0, 1, 0],
    [0, 0, 1, 0, 0],
    [0, 1, 0, 0, 0],
    [1, 0, 0, 0, 0],
  ],
]

export const GAME_MAPS = [
  bigBox,
  smallBox,
  bigCorners,
  letterX,
  letterT,
  cross,
  letterL,
  coverAll,
]
