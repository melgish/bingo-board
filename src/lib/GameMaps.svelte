<script lang="ts">
  import { onMount } from "svelte"
  import GameMap from "./GameMap.svelte"
  import { GAME_MAPS, normals } from "./game-maps"

  let active: (typeof normals)|(typeof normals[0]) = GAME_MAPS[0]
  let n = 0

  const advanceNormals = () => {
    if (active === normals) {
      n = (n + 1) % normals.length
    }
  }

  onMount(() => {
    // Animate normal game every 1/2 second
    const i = setInterval(advanceNormals, 500)
    return () => clearInterval(i)
  })
</script>

<div class="games" data-testid="games">
  <div class:active={normals == active}>
    <GameMap game={normals[n]} on:click={() => (active = normals)} />
  </div>
  {#each GAME_MAPS as game}
    <div class:active={game == active}>
      <GameMap {game} on:click={() => (active = game)} />
    </div>
  {/each}
</div>

<style>
  .games {
    display: flex;
    flex-direction: row;
    margin-bottom: 0.25rem;
  }
  .games > div {
    opacity: 0.5;
    margin-right: 0.5rem;
  }
  .games > div.active {
    opacity: 1;
  }
</style>
