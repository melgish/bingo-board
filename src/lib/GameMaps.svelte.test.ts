import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { act, fireEvent, render, screen } from "@testing-library/svelte"
import GameMaps from "./GameMaps.svelte"
import { GAME_MAPS } from "./game-maps"

function getButton(index: number)  {
  return screen.getAllByRole("button")[index];
}

// Take control of timers to control animations.
beforeAll(() => {
  vi.useFakeTimers()
})
afterAll(() => {
  vi.useRealTimers()
})

describe("GameMaps", () => {
  describe("when rendered", () => {
    it("should display all maps", async () => {
      render(GameMaps);
      const games = screen.getAllByRole("button")

      // Should have 1 for each map, plus normals
      expect(games.length).toBe(GAME_MAPS.length + 1)
    })
  })

  describe("when interval passes", () => {
    it("should not update normal to next map", async () => {
      render(GameMaps);

      expect(getButton(0)).toMatchSnapshot()

      await act(() => vi.runOnlyPendingTimers())
      await act(() => vi.runOnlyPendingTimers())

      expect(getButton(0)).toMatchSnapshot()
    })
  })

  describe("when normal game is selected", () => {
    it("should animate the map", async () => {
      render(GameMaps);

      await fireEvent.click(getButton(0));

      await act(() => vi.runOnlyPendingTimers())
      await act(() => vi.runOnlyPendingTimers())

      expect(getButton(0)).toMatchSnapshot()
    })
  })
})
