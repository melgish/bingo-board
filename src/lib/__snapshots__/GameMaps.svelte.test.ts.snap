// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GameMaps > when interval passes > should not update normal to next map 1`] = `
<button
  aria-label="Select Game"
  class="s-Foe4RCTFnPEn"
  type="button"
>
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
</button>
`;

exports[`GameMaps > when interval passes > should not update normal to next map 2`] = `
<button
  aria-label="Select Game"
  class="s-Foe4RCTFnPEn"
  type="button"
>
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
</button>
`;

exports[`GameMaps > when normal game is selected > should animate the map 1`] = `
<button
  aria-label="Select Game"
  class="s-Foe4RCTFnPEn"
  type="button"
>
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  <div
    class="s-Foe4RCTFnPEn filled"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  <div
    class="s-Foe4RCTFnPEn"
  />
  
</button>
`;
